geode-wan/src/test/java/org/apache/geode/cache/client/internal/locator/wan/LocatorMembershipListenerTest.java

```diff
@@ -0,0 +1,313 @@
+    allLocatorsInfo.put(2,
+        new HashSet<>(Arrays.asList(distributionLocator1Site2, distributionLocator2Site2)));
+    allLocatorsInfo.put(3, new HashSet<>(Arrays.asList(distributionLocator1Site3,
+        distributionLocator2Site3, distributionLocator3Site3)));
+    when(locatorMembershipListener.getAllLocatorsInfo()).thenReturn(allLocatorsInfo);
+
+    locatorMembershipListener.locatorJoined(1, joiningLocator, distributionLocator1Site1);
+    joinLocatorsDistributorThread();
+
+    verify(tcpClient).requestToServer(distributionLocator3Site1.getHost(),
+        new LocatorJoinMessage(1, joiningLocator, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(joiningLocator.getHost(),
+        new LocatorJoinMessage(1, distributionLocator3Site1, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(distributionLocator1Site2.getHost(),
+        new LocatorJoinMessage(1, joiningLocator, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(joiningLocator.getHost(),
+        new LocatorJoinMessage(2, distributionLocator1Site2, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(distributionLocator2Site2.getHost(),
+        new LocatorJoinMessage(1, joiningLocator, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(joiningLocator.getHost(),
+        new LocatorJoinMessage(2, distributionLocator2Site2, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(distributionLocator1Site3.getHost(),
+        new LocatorJoinMessage(1, joiningLocator, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(joiningLocator.getHost(),
+        new LocatorJoinMessage(3, distributionLocator1Site3, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(distributionLocator2Site3.getHost(),
+        new LocatorJoinMessage(1, joiningLocator, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(joiningLocator.getHost(),
+        new LocatorJoinMessage(3, distributionLocator2Site3, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(distributionLocator3Site3.getHost(),
+        new LocatorJoinMessage(1, joiningLocator, distributionLocator1Site1, ""),
+        DistributionConfig.DEFAULT_MEMBER_TIMEOUT, false);
+    verify(tcpClient).requestToServer(joiningLocator.getHost(),
```

Comment:
One of the unit testing best practices is `one assertion per test`. These tests are performing and verifying an awful lot within each test. I can recommend some good references if you want them.

One example (just the 1st page I found with a google search): https://stackify.com/unit-testing-basics-best-practices/. Search down for Unit Testing Best Practices and the author provides some detail for each of these:
1. Arrange, Act, Assert
2. **One Assert Per Test Method**
3. Avoid Test Interdependence
4. Keep It Short, Sweet, and Visible
5. Recognize Test Setup Pain as a Smell
6. Add Them to the Build

Reply:
I don't fully agree with this one but I'll improve the test methods anyway.
Even though there seems to be more than one assertion per test and I could simplify all of them with just one call to `verify(expectedTotalAmount)`, the test is actually verifying **one single method on the class under test**. Also, the test itself aims to verify that every expected combination of messages has been sent, that's why I do want to keep multiple assertions.